pkg/auth/providers/saml/saml_provider.go

```diff
@@ -0,0 +1,237 @@
+package saml
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/crewjam/saml"
+	"github.com/crewjam/saml/samlsp"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/auth/providers/common"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/client/management/v3"
+	publicclient "github.com/rancher/types/client/management/v3public"
+	"github.com/rancher/types/config"
+	"github.com/rancher/types/user"
+	"github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const PingName = "ping"
+
+type Provider struct {
+	Ctx             context.Context
+	AuthConfigs     v3.AuthConfigInterface
+	UserMGR         user.Manager
+	TokenMGR        *tokens.Manager
+	ServiceProvider *saml.ServiceProvider
+	Name            string
+	UserType        string
+	GroupType       string
+	ClientState     samlsp.ClientState
+}
+
+var SamlProviders = make(map[string]*Provider)
+
+func Configure(ctx context.Context, mgmtCtx *config.ScaledContext, userMGR user.Manager, tokenMGR *tokens.Manager, name string) common.AuthProvider {
+	samlp := &Provider{
+		Ctx:         ctx,
+		AuthConfigs: mgmtCtx.Management.AuthConfigs(""),
+		UserMGR:     userMGR,
+		TokenMGR:    tokenMGR,
+		Name:        name,
+		UserType:    name + "_user",
+		GroupType:   name + "_group",
+	}
+	SamlProviders[PingName] = samlp
+	return samlp
+}
+
+func (s *Provider) GetName() string {
+	return s.Name
+}
+
+func (s *Provider) CustomizeSchema(schema *types.Schema) {
+	schema.ActionHandler = s.actionHandler
+	schema.Formatter = s.formatter
+}
+
+func (s *Provider) TransformToAuthProvider(authConfig map[string]interface{}) map[string]interface{} {
+	p := common.TransformToAuthProvider(authConfig)
+	if s.Name == PingName {
```

Comment:
Is the logic going to vary from provider to provider? seems like it should be the same logic and so we should not need if blocks for each provider. If the only difference is the name of the constant we use, I think we can just safely use the same constant for all providers

Reply:
We're also using the redirectURL from different providers, Ping vs Shibboleth, hence the if-else

Reply:
ok, cool